File: 0206-reverse-linked-list/0206-reverse-linked-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var reverseList = function(head) {
    var previousNode = null;
    var currentNode = head;
    
    while (currentNode) {
        var nextNode = currentNode.next;
        currentNode.next = previousNode;
        previousNode = currentNode;
        currentNode = nextNode;
        // console.log(`previousNode: ${previousNode.val}`)
    }
    
    return previousNode;
};

File: 0206-reverse-linked-list/0206-reverse-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head:
            return None
        
        prev_node = None
        curr_node = head
        next_node = head.next
        
        while next_node:
            next_next_node = next_node.next
            next_node.next = curr_node
            curr_node.next = prev_node
            prev_node = curr_node
            curr_node = next_node
            next_node = next_next_node
        
        return curr_node